frontend/z3_axioms.py

```diff
@@ -1,203 +1,404 @@
-            Exists([x], And(sliced == List(x), value == List(x))))
+        And(types.subtype(lower, types.int), types.subtype(upper, types.int), types.subtype(step, types.int),
+            Exists([x], And(sliced == types.list(x), value == types.list(x))))
     ]
 
 
-def delete_subscript(indexed):
+def delete_subscript(indexed, types):
+    """Constraints for subscript deletion
+    
+    Prevent subscript deletion of tuples, strings and bytes (Immutable sequences)
+    """
     return [
         Not(Or(
-            indexed == String,
-            indexed == Bytes,
-            subtype(indexed, Tuple)
+            indexed == types.string,
+            indexed == types.bytes,
+            types.subtype(indexed, types.tuple)
         ))
     ]
 
 
-def body(result, new):
+def body(result, new, types):
+    """Constraints for body statements
+    
+    The body type is the super-type of all its statements, or none if no statement returns type.
+    """
     return [
-        Implies(new != zNone, subtype(new, result))
+        Implies(new != types.none, types.subtype(new, result))
     ]
 
 
-def control_flow(body, orelse, result):
+def control_flow(then, orelse, result, types):
+    """Constraints for control-flow blocks (if/else, while, for)"""
     # TODO numeric casting
     return [
-        Implies(orelse == zNone, result == body),
-        Implies(orelse != zNone, And(
-            subtype(body, result),
-            subtype(orelse, result)
+        Implies(orelse == types.none, result == then),
+        Implies(orelse != types.none, And(
+            types.subtype(then, result),
+            types.subtype(orelse, result)
         ))
     ]
 
 
-def for_loop(iterable, target):
+def for_loop(iterable, target, types):
+    """Constraints for for-loop iterable and iteration target"""
+    x = Const("x", types.type_sort)
     return [
         Or(
-            iterable == List(target),
-            iterable == Set(target),
-            Exists([x], iterable == Dict(target, x), patterns=[Dict(target, x)]),
-            And(iterable == String, target == String),
-            And(iterable == Bytes, target == Bytes)
+            iterable == types.list(target),
+            iterable == types.set(target),
+            Exists([x], iterable == types.dict(target, x), patterns=[types.dict(target, x)]),
+            And(iterable == types.string, target == types.string),
+            And(iterable == types.bytes, target == types.bytes)
         )
     ]
 
 
-def try_except(body, orelse, final, result):
+def try_except(then, orelse, final, result, types):
+    """Constraints for try/except block"""
     return [
-        subtype(body, result),
-        subtype(orelse, result),
-        subtype(final, result)
+        types.subtype(then, result),
+        types.subtype(orelse, result),
+        types.subtype(final, result)
     ]
+
+
+def instance_axioms(called, args, result, types):
+    """Constraints for class instantiation"""
```

Comment:
A class instantiation corresponds to a normal function call to the `__init__` function, where
the return type will be an instance of this class.
    
The called maybe of any user-defined type in the program, so the call is asserted
with the `__init__` function of every call.

Added more comments.